torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -0,0 +1,117 @@
+  xla::int64 rank_delta = to_shape.rank() - op_shape.rank();
+  XLA_CHECK_GT(rank_delta, 0);
+
+  std::vector<xla::int64> reshape_sizes(to_shape.rank(), 1);
+  std::copy(op_shape.dimensions().begin(), op_shape.dimensions().end(),
+            reshape_sizes.begin() + rank_delta);
+  xla::XlaOp result = xla::Reshape(op, reshape_sizes);
+
+  std::vector<xla::int64> bcast_sizes(
+      to_shape.dimensions().begin(),
+      to_shape.dimensions().begin() + rank_delta);
+  bcast_sizes.insert(bcast_sizes.end(), op_shape.dimensions().begin(),
+                     op_shape.dimensions().end());
+  return std::make_pair(
+      xla::BroadcastInDim(result, bcast_sizes,
+                          xla::util::Iota<xla::int64>(to_shape.rank())),
+      xla::ShapeUtil::MakeShape(op_shape.element_type(), bcast_sizes));
+}
+
+std::pair<xla::XlaOp, xla::XlaOp> DotBroadcast(const xla::XlaOp& lhs,
+                                               const xla::Shape& lhs_shape,
+                                               const xla::XlaOp& rhs,
+                                               const xla::Shape& rhs_shape) {
+  auto lhs_dimensions = lhs_shape.dimensions();
+  auto rhs_dimensions = rhs_shape.dimensions();
+  XLA_CHECK_EQ(lhs_dimensions.size(), rhs_dimensions.size());
+  for (xla::int64 i = 0; i < lhs_dimensions.size() - 2; ++i) {
+    if (lhs_dimensions[i] == rhs_dimensions[i]) {
+      continue;
+    }
+    if (lhs_dimensions[i] == 1) {
+      lhs_dimensions[i] = rhs_dimensions[i];
+    } else if (rhs_dimensions[i] == 1) {
+      rhs_dimensions[i] = lhs_dimensions[i];
+    } else {
+      XLA_ERROR() << "Unsupported DotBroadcast: " << lhs_shape << " vs. "
+                  << rhs_shape;
+    }
+  }
+
+  xla::XlaOp blhs = lhs;
+  xla::XlaOp brhs = rhs;
+  if (lhs_dimensions != lhs_shape.dimensions()) {
+    blhs =
+        xla::BroadcastInDim(lhs, lhs_dimensions,
+                            xla::util::Iota<xla::int64>(lhs_dimensions.size()));
+  }
+  if (rhs_dimensions != rhs_shape.dimensions()) {
+    brhs =
+        xla::BroadcastInDim(rhs, rhs_dimensions,
+                            xla::util::Iota<xla::int64>(rhs_dimensions.size()));
+  }
+  return std::make_pair(blhs, brhs);
+}
+
+}  // namespace
+
+xla::XlaOp CreateMatMul(const xla::XlaOp& lhs, const xla::XlaOp& rhs) {
+  // Expand cases in https://pytorch.org/docs/stable/torch.html#torch.matmul
+  xla::Shape lhs_shape = XlaHelpers::ShapeOfXlaOp(lhs);
+  xla::Shape rhs_shape = XlaHelpers::ShapeOfXlaOp(rhs);
+  if ((lhs_shape.rank() == 1 && rhs_shape.rank() == 1) ||
+      (lhs_shape.rank() == 2 && rhs_shape.rank() == 2) ||
+      (lhs_shape.rank() == 2 && rhs_shape.rank() == 1)) {
+    return xla::Dot(lhs, rhs);
+  }
+  if (lhs_shape.rank() == 1 && rhs_shape.rank() == 2) {
+    xla::XlaOp reshaped_lhs = xla::Reshape(lhs, {1, lhs_shape.dimensions(0)});
+    return xla::Reshape(xla::Dot(reshaped_lhs, rhs), {rhs_shape.dimensions(1)});
+  }
+  if (lhs_shape.rank() >= 1 && rhs_shape.rank() >= 1 &&
```

Comment:
I assume you've looked into `xla::BatchDot` and doesn't do what we need, correct?

Reply:
Yep, does not do all the reshape and broadcast required by PT.